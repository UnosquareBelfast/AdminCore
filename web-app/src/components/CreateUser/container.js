import React, { Component } from 'react';
import moment from 'moment';
import employeeStatus from '../../utilities/employeeStatus';
import { createUser } from '../../services/userService';

import {
  getFormElementsArray,
  updateFormDataOnChange,
  isFormValidOnChange,
  getFormDataOnSubmit,
} from '../../utilities/forms';

export default Wrapped =>
  class extends Component {
    constructor(props) {
      super(props);
      this.state = {
        form: this.initialFormState(),
        formIsValid: false,
        success: null,
        error: null,
        loading: false,
      };
    }

    initialFormState() {
      return {
        forename: {
          label: 'Forename:',
          elementType: 'input',
          elementConfig: {
            type: 'text',
            placeholder: 'Enter your forename',
          },
          value: '',
          validation: {
            required: true,
          },
          valid: false,
          touched: false,
        },
        surname: {
          label: 'Surname:',
          elementType: 'input',
          elementConfig: {
            type: 'text',
            placeholder: 'Enter your surname',
          },
          value: '',
          validation: {
            required: true,
          },
          valid: false,
          touched: false,
        },
        email: {
          label: 'Email:',
          elementType: 'input',
          elementConfig: {
            type: 'email',
            placeholder: 'Enter your email address',
          },
          value: '',
          validation: {
            required: true,
            isEmail: true,
          },
          valid: false,
          touched: false,
        },
        password: {
          label: 'Password:',
          elementType: 'input',
          elementConfig: {
            type: 'password',
            placeholder: 'Enter a password',
          },
          value: '',
          validation: {
            required: true,
            minLength: 6,
          },
          valid: false,
          touched: false,
        },
        country: {
          label: 'Country:',
          elementType: 'select',
          elementConfig: {
            options: [
              { value: 1, displayValue: 'Northern Ireland' },
              { value: 2, displayValue: 'Mexico' },
            ],
          },
          value: 1,
          validation: {},
          valid: true,
        },
        status: {
          //TODO Create enum file employeeStatus.ACTIVE
          label: 'Status:',
          elementType: 'select',
          elementConfig: {
            options: [
              { value: employeeStatus.ACTIVE, displayValue: 'Active' },
              { value: employeeStatus.INACTIVE, displayValue: 'Inactive' },
            ],
          },
          value: 1,
          validation: {},
          valid: true,
        },
        employeeRole: {
          label: 'Employee Role:',
          elementType: 'select',
          elementConfig: {
            options: [
              { value: 1, displayValue: 'Team Leader' },
              { value: 2, displayValue: 'System Admin' },
              { value: 3, displayValue: 'Employee' },
            ],
          },
          value: 3,
          validation: {},
          valid: true,
        },
        startDate: {
          label: 'Start Date:',
          elementType: 'date',
          elementConfig: {
            type: 'text',
            placeholder: 'Enter a start date',
          },
          value: moment(),
          validation: {},
          valid: true,
        },
      };
    }

    handleFormChange = (event, inputIdentifier) => {
      const inputValue =
        inputIdentifier !== 'startDate' ? event.target.value : event;

      const updatedForm = updateFormDataOnChange(
        this.state.form,
        inputIdentifier,
        inputValue,
      );

      let formIsValid = isFormValidOnChange(updatedForm);
      this.setState({ form: updatedForm, formIsValid: formIsValid });
    };

    handleFormSubmit = e => {
      e.preventDefault();
      const formData = getFormDataOnSubmit(this.state.form);

      formData.startDate = formData.startDate.toISOString();
      this.setState({ loading: true });
      return createUser(formData)
        .then(() => {
          this.setState({
            loading: false,
            success: true,
            form: this.initialFormState(),
          });
        })
        .catch(error => {
          this.setState({ error, loading: false });
        });
    };

    render() {
      return (
        <Wrapped
          formElementsArray={getFormElementsArray(this.state.form)}
          formIsValid={this.state.formIsValid}
          submitForm={this.handleFormSubmit}
          formChanged={(event, id) => this.handleFormChange(event, id)}
          success={this.state.success}
          error={this.state.error}
          loading={this.state.loading}
        />
      );
    }
  };
