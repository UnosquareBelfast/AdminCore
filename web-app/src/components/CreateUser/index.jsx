import React, { Fragment } from 'react';
import { PropTypes as PT } from 'prop-types';
import CreateUserForm from './CreateUserForm';
import { Button } from '../common';
import { CornerButton } from '../common_styled';

export const CreateUser = ({ history }) => {
  return (
    <Fragment>
      <CornerButton>
        <Button
          onClick={() => history.replace('/admin/employees')}
          label="View Employees"
        />
      </CornerButton>

      <h2>Create User</h2>
      <CreateUserForm />
    </Fragment>
  );
};

CreateUser.propTypes = {
  history: PT.object.isRequired,
};

export default CreateUser;
