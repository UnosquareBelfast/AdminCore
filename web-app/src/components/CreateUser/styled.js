import styled from 'styled-components';

export const Container = styled.div`
  label {
    display: block;
  }
  input,
  select {
    display: block;
    margin: 2px 0 10px 0;
    width: 90%;
    height: 15px;
    padding: 0 5px;
  }
`;
