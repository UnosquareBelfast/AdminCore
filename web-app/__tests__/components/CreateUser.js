import React from 'react';
import moment from 'moment';
import { shallow } from 'enzyme';
import { CreateUser } from '../../src/components/CreateUser/index';

const formData = {
  forename: '',
  surname: '',
  email: 'email@email.com',
  password: 'password',
  country: 'Northern Ireland',
  employeeRole: 'Admin',
  startDate: moment(),
  endDate: moment(),
};

describe('BookingModal', () => {
  it('renders correctly', () => {
    const formChanged = jest.fn();
    const startDateChanged = jest.fn();
    const submitForm = jest.fn();

    const wrapper = shallow(<CreateUser 
      formData={formData}
      formChanged={formChanged}
      startDateChanged={startDateChanged}
      error={null}
      loading={false}
      submitForm={ submitForm }/>);

    //Forename
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.find('#forenameLbl').text()).toEqual('Forename:');
    expect(wrapper.find('#forenameLbl').childAt(1).type()).toEqual('input');        
    expect(wrapper.find('#forenameLbl').childAt(1).prop('type')).toEqual('text');
    expect(wrapper.find('#forenameLbl').childAt(1).prop('name')).toEqual('forename');
    expect(wrapper.find('#forenameLbl').childAt(1).prop('value')).toEqual(formData.forename);
    expect(wrapper.find('#forenameLbl').childAt(1).prop('onChange')).toEqual(formChanged);
    expect(wrapper.find('#forenameLbl').childAt(1).prop('disabled')).toEqual(false);

    //Surname
    expect(wrapper.find('#surnameLbl').text()).toEqual('Surname:');
    expect(wrapper.find('#surnameLbl').childAt(1).type()).toEqual('input');        
    expect(wrapper.find('#surnameLbl').childAt(1).prop('type')).toEqual('text');
    expect(wrapper.find('#surnameLbl').childAt(1).prop('name')).toEqual('surname');
    expect(wrapper.find('#surnameLbl').childAt(1).prop('value')).toEqual(formData.surname);
    expect(wrapper.find('#surnameLbl').childAt(1).prop('onChange')).toEqual(formChanged);
    expect(wrapper.find('#surnameLbl').childAt(1).prop('disabled')).toEqual(false);

    //Email
    expect(wrapper.find('#emailLbl').text()).toEqual('Email:');
    expect(wrapper.find('#emailLbl').childAt(1).type()).toEqual('input');        
    expect(wrapper.find('#emailLbl').childAt(1).prop('type')).toEqual('text');
    expect(wrapper.find('#emailLbl').childAt(1).prop('name')).toEqual('email');
    expect(wrapper.find('#emailLbl').childAt(1).prop('value')).toEqual(formData.email);
    expect(wrapper.find('#emailLbl').childAt(1).prop('onChange')).toEqual(formChanged);
    expect(wrapper.find('#emailLbl').childAt(1).prop('disabled')).toEqual(false);

    //Password
    expect(wrapper.find('#passwordLbl').text()).toEqual('Password:');
    expect(wrapper.find('#passwordLbl').childAt(1).type()).toEqual('input');    
    expect(wrapper.find('#passwordLbl').childAt(1).prop('type')).toEqual('password');
    expect(wrapper.find('#passwordLbl').childAt(1).prop('name')).toEqual('password');
    expect(wrapper.find('#passwordLbl').childAt(1).prop('value')).toEqual(formData.password);
    expect(wrapper.find('#passwordLbl').childAt(1).prop('autoComplete')).toEqual('new-password');
    expect(wrapper.find('#passwordLbl').childAt(1).prop('onChange')).toEqual(formChanged);
    expect(wrapper.find('#passwordLbl').childAt(1).prop('disabled')).toEqual(false);

    //Country
    expect(wrapper.find('#countryLbl').childAt(1).type()).toEqual('select');
    expect(wrapper.find('#countryLbl').childAt(1).prop('value')).toEqual(formData.country);

    expect(wrapper.find('#countryLbl').childAt(1).childAt(0).type()).toEqual('option');
    expect(wrapper.find('#countryLbl').childAt(1).childAt(0).prop('value')).toEqual(1);
    expect(wrapper.find('#countryLbl').childAt(1).childAt(0).text()).toEqual('Northern Ireland');
    expect(wrapper.find('#countryLbl').childAt(1).childAt(1).type()).toEqual('option');
    expect(wrapper.find('#countryLbl').childAt(1).childAt(1).prop('value')).toEqual(2);
    expect(wrapper.find('#countryLbl').childAt(1).childAt(1).text()).toEqual('Mexico');
    
    //Role
    expect(wrapper.find('#roleLbl').childAt(1).type()).toEqual('select');
    expect(wrapper.find('#roleLbl').childAt(1).prop('value')).toEqual(formData.employeeRole);
    expect(wrapper.find('#roleLbl').childAt(1).prop('onChange')).toEqual(formChanged);
    expect(wrapper.find('#roleLbl').childAt(1).prop('disabled')).toBe(false);

    expect(wrapper.find('#roleLbl').childAt(1).childAt(0).type()).toEqual('option');
    expect(wrapper.find('#roleLbl').childAt(1).childAt(0).prop('value')).toEqual(3);
    expect(wrapper.find('#roleLbl').childAt(1).childAt(0).text()).toEqual('Employee');
    expect(wrapper.find('#roleLbl').childAt(1).childAt(1).type()).toEqual('option');
    expect(wrapper.find('#roleLbl').childAt(1).childAt(1).prop('value')).toEqual(1);
    expect(wrapper.find('#roleLbl').childAt(1).childAt(1).text()).toEqual('Team Leader');
    expect(wrapper.find('#roleLbl').childAt(1).childAt(2).type()).toEqual('option');
    expect(wrapper.find('#roleLbl').childAt(1).childAt(2).prop('value')).toEqual(2);
    expect(wrapper.find('#roleLbl').childAt(1).childAt(2).text()).toEqual('System Admin');

    //Start Date
    const text = wrapper.find('#startDateLbl').text().includes('Start Date:');
    expect(text).toBe(true);    
    expect(wrapper.find('#startDateLbl').childAt(1).prop('selected')).toEqual(formData.startDate);
    expect(wrapper.find('#startDateLbl').childAt(1).prop('onChange')).toEqual(startDateChanged);
    expect(wrapper.find('#startDateLbl').childAt(1).prop('disabled')).toEqual(false);

    expect(wrapper.find('#createUserBtn').exists()).toBe(true);
    expect(wrapper.find('#createUserBtn').prop('label')).toEqual('Create User');
    expect(wrapper.find('#createUserBtn').prop('onClick')).toEqual(submitForm);
    expect(wrapper.find('#createUserBtn').prop('disabled')).toBe(false);

    expect(wrapper.find('#userCreatedSuccess').exists()).toBe(false);
    expect(wrapper.find('#errorCreateUser').exists()).toBe(true);

  });
  it('renders correctly when error is passed in and success is true', () => {
    const formChanged = jest.fn();
    const startDateChanged = jest.fn();
    const submitForm = jest.fn();

    const wrapper = shallow(<CreateUser 
      formData={formData}
      formChanged={formChanged}
      startDateChanged={startDateChanged}
      error={ {} }
      success
      submitForm={ submitForm }/>);

    expect(wrapper.find('#userCreatedSuccess').exists()).toBe(true);
    expect(wrapper.find('#errorCreateUser').exists()).toBe(true);
  });
  it('all fields are disabled when loading is true', () => {
    const formChanged = jest.fn();
    const startDateChanged = jest.fn();
    const submitForm = jest.fn();

    const wrapper = shallow(<CreateUser 
      formData={formData}
      formChanged={formChanged}
      startDateChanged={startDateChanged}
      error={ {} }
      success
      loading
      submitForm={ submitForm }/>);

    wrapper.find('input').forEach((node) => {
      expect(node.prop('disabled')).toBe(true);
    });
    wrapper.find('select').forEach((node) => {
      expect(node.prop('disabled')).toBe(true);
    });
  });
});
