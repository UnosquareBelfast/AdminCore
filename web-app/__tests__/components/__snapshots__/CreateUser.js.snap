// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingModal renders correctly 1`] = `
<Card>
  <styled.div>
    <h3>
      Create Employee
    </h3>
    <form
      autoComplete="off"
    >
      <label
        id="forenameLbl"
      >
        Forename:
        <input
          disabled={false}
          name="forename"
          onChange={[MockFunction]}
          type="text"
          value=""
        />
      </label>
      <label
        id="surnameLbl"
      >
        Surname:
        <input
          disabled={false}
          name="surname"
          onChange={[MockFunction]}
          type="text"
          value=""
        />
      </label>
      <label
        id="emailLbl"
      >
        Email:
        <input
          disabled={false}
          name="email"
          onChange={[MockFunction]}
          type="text"
          value="email@email.com"
        />
      </label>
      <label
        id="passwordLbl"
      >
        Password:
        <input
          autoComplete="new-password"
          disabled={false}
          name="password"
          onChange={[MockFunction]}
          type="password"
          value="password"
        />
      </label>
      <label
        id="countryLbl"
      >
        Country:
        <select
          disabled={false}
          name="country"
          onChange={[MockFunction]}
          value="Northern Ireland"
        >
          <option
            value={1}
          >
            Northern Ireland
          </option>
          <option
            value={2}
          >
            Mexico
          </option>
        </select>
      </label>
      <label
        id="roleLbl"
      >
        Role:
        <select
          disabled={false}
          name="employeeRole"
          onChange={[MockFunction]}
          value="Admin"
        >
          <option
            value={3}
          >
            Employee
          </option>
          <option
            value={1}
          >
            Team Leader
          </option>
          <option
            value={2}
          >
            System Admin
          </option>
        </select>
      </label>
      <label
        id="startDateLbl"
      >
        Start Date:
        <DatePicker
          allowSameDay={false}
          dateFormat="L"
          dateFormatCalendar="MMMM YYYY"
          disabled={false}
          disabledKeyboardNavigation={false}
          dropdownMode="scroll"
          monthsShown={1}
          onBlur={[Function]}
          onChange={[MockFunction]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          onYearChange={[Function]}
          preventOpenOnFocus={false}
          selected={"2018-01-01T09:00:00.000Z"}
          shouldCloseOnSelect={true}
          showTimeSelect={false}
          timeCaption="Time"
          timeIntervals={30}
          withPortal={false}
        />
      </label>
      <Button
        disabled={false}
        id="createUserBtn"
        label="Create User"
        onClick={[MockFunction]}
      />
    </form>
    <Errorbox
      error={null}
      id="errorCreateUser"
      label="Error creating user"
    />
  </styled.div>
</Card>
`;
